interface/src/CrashHandler_Crashpad.cpp

```diff
@@ -70,8 +76,99 @@ LONG WINAPI vectoredExceptionHandler(PEXCEPTION_POINTERS pExceptionInfo) {
     }
 
+    if (pExceptionInfo->ExceptionRecord->ExceptionCode == 0xE06D7363) {
+        fatalCxxException(pExceptionInfo);
+        client->DumpAndCrash(pExceptionInfo);
+    }
+
     return EXCEPTION_CONTINUE_SEARCH;
 }
+
+#pragma pack(push, ehdata, 4)
+
+struct PMD_internal {
+    int mdisp;
+    int pdisp;
+    int vdisp;
+};
+
+struct ThrowInfo_internal {
+    __int32 attributes;
+    __int32 pmfnUnwind;           // 32-bit RVA
+    __int32 pForwardCompat;       // 32-bit RVA
+    __int32 pCatchableTypeArray;  // 32-bit RVA
+};
+
+struct CatchableType_internal {
+    __int32 properties;
+    __int32 pType;                // 32-bit RVA
+    PMD_internal thisDisplacement;
+    __int32 sizeOrOffset;
+    __int32 copyFunction;         // 32-bit RVA
+};
+
+#pragma warning(disable : 4200)
+struct CatchableTypeArray_internal {
+    int nCatchableTypes;
+    __int32 arrayOfCatchableTypes[0]; // 32-bit RVA
+};
+#pragma warning(default : 4200)
+
+#pragma pack(pop, ehdata)
+
+// everything inside this function is extremely undocumented, attempting to extract
+// the underlying C++ exception type (or at least its name) before throwing the whole
+// mess at crashpad
+void fatalCxxException(PEXCEPTION_POINTERS pExceptionInfo) {
+    PEXCEPTION_RECORD ExceptionRecord = pExceptionInfo->ExceptionRecord;
+
+    if(ExceptionRecord->NumberParameters != 4 || ExceptionRecord->ExceptionInformation[0] != 0x19930520) {
+        // doesn't match expected parameter counts or magic numbers
+        return;
+    }
+
+    //ULONG_PTR signature = ExceptionRecord->ExceptionInformation[0];
+    void* pExceptionObject = reinterpret_cast<void*>(ExceptionRecord->ExceptionInformation[1]); // the object that generated the exception
+    ThrowInfo_internal* pThrowInfo = reinterpret_cast<ThrowInfo_internal*>(ExceptionRecord->ExceptionInformation[2]);
+    ULONG_PTR moduleBase = ExceptionRecord->ExceptionInformation[3];
+    if(moduleBase == 0 || pThrowInfo == NULL) {
+      return; // broken assumption
+    }
+
+    // now we start breaking the pThrowInfo internal structure apart
+    if(pThrowInfo->pCatchableTypeArray == 0) {
+      return; // broken assumption
+    }
+    CatchableTypeArray_internal* pCatchableTypeArray = reinterpret_cast<CatchableTypeArray_internal*>(moduleBase + pThrowInfo->pCatchableTypeArray);
+    if(pCatchableTypeArray->nCatchableTypes == 0 || pCatchableTypeArray->arrayOfCatchableTypes[0] == 0) {
+      return; // broken assumption
+    }
+    CatchableType_internal* pCatchableType = reinterpret_cast<CatchableType_internal*>(moduleBase + pCatchableTypeArray->arrayOfCatchableTypes[0]);
+    if(pCatchableType->pType == 0) {
+      return; // broken assumption
+    }
+    const std::type_info* type = reinterpret_cast<std::type_info*>(moduleBase + pCatchableType->pType);
+
+    // we're crashing, not really sure it matters who's currently holding the lock (although this could in theory really mess things up
+    annotationMutex.try_lock();
```

Comment:
Return code isn't checked. Mutex isn't ever unlocked.

This seems very dubious. If the idea here is that since we're crashing, the mutex may not ever be unlocked, perhaps we should spin and wait here for some suitable timeout instead, and then abort outright if it doesn't work? A corrupt report is unhelpful anyway.

I'd like to hear a good reason for proceeding without taking the lock. Is it likely to happen? Is the resulting bug report going to be useful?

Reply:
The only contention here would be if a different thread was trying to set an annotation while the system was crashing.  I would have no problem with a spinlock with a timeout, but considering the system is in a degraded state I would want a timeout with "proceed anyhow without lock".

I'd appreciate suggestions here for improving this.

Reply:
Mm. I'll admit that I'm a bit out of my depth here, but this PR is for exception handling, so probably the application is otherwise in an usable state, and all the other threads are running, right? So all existing threads should be proceeding forward without trouble?

Also, if a thread stops with the mutex held, the likelihood of it corrupting the crashpad_info state is high, and if that happens, we would seem to be screwed anyway.


But (very rough idea here), how about an atomic swap scheme?

Reply:
Atomic swap wouldn't really work as this mutex is protecting crashpadAnnotations which is a "crashpad::SimpleStringDictionary"

Reply:
I'm kinda arguing in my brain about this, yes it's unlikely we could deadlock inside the crash logic waiting for the mutex but... it does seem to be a possibility.  At the moment I think I'd prefer a spinlock with a timeout of like a couple seconds (the lock really shouldn't be getting held that long)

Reply:
Hmm, there is a second, more complex way of adding annotations to crashpad that we aren't currently using; I could look into using that for this which may bypass the need for locking the map altogether (as we wouldn't be using it)

Reply:
The rough idea I have is:

1. Have a crashpad::SimpleStringDictionary* current_annotations
2. Take the lock.
3. Create a new crashpad::SimpleStringDictionary
4. Copy the entirety of what's in current_annotations
5. Apply the change.
6. Atomically overwrite current_annotations with the pointer to the new, modified dictionary.
7. Unlock.

Then the crash code can go by whatever is in current_annotations, without having to take a lock.

Reply:
Can't change the pointer to annotations or move it around memory like that I don't think, it's part of the glue that connects this to the external crashpad executable.

There is a second type of annotation that doesn't use this map that I may be able to use.  Registering it with the external executable involves creating a linked list.  Nowhere else in the program is using the second methodology so we may not even need to take out a mutex (If the client code were properly written then THIS would be a good place for atomic variables)

Reply:
Hmm. I poked at it, and realized that crashpad_info->set_simple_annotations(crashpadAnnotations) is called in setCrashAnnotation, so this whole scheme I had in mind may not work.

But upon further thought, I still don't think ignoring the mutex is going to be any good. It's needed to protect Crashpad's own data, and if that's corrupt then it's unlikely to do much good. It'll either crash there, or produced a garbled report in the best case, and that doesn't seem like an useful thing.

Plus, you're talking about exception handling here, and an exception shouldn't involve such dramatic malfunction that other threads suddenly grind to a halt, right?